massa-models/src/denunciation.rs

```diff
@@ -0,0 +1,685 @@
+    index: u32,
+    hash_1: Hash,
+    hash_2: Hash,
+    signature_1: Signature,
+    signature_2: Signature,
+}
+
+impl EndorsementDenunciation {
+    /// Compute hash for Endorsement content
+    fn compute_content_hash(content: &Endorsement) -> Result<Hash, SerializeError> {
+        let mut buf = Vec::new();
+        let endorsement_serializer = EndorsementSerializer::new();
+        endorsement_serializer.serialize(content, &mut buf)?;
+        Ok(Hash::compute_from(&buf))
+    }
+
+    /// Rebuild full hash of SecureShareEndorsement from given arguments
+    fn compute_hash_for_sig_verif(
+        public_key: &PublicKey,
+        slot: &Slot,
+        index: &u32,
+        content_hash: &Hash,
+    ) -> Result<Hash, SerializeError> {
+        let mut hash_data = Vec::new();
+        let mut buf = Vec::new();
+        let de_data_serializer = DenunciationDataSerializer::new();
+
+        // Public key
+        hash_data.extend(public_key.to_bytes());
+
+        // Ser slot & index
+        let denunciation_data = DenunciationData::Endorsement((*slot, *index));
+        de_data_serializer.serialize(&denunciation_data, &mut buf)?;
+        hash_data.extend(&buf);
+        buf.clear();
+
+        // Add content hash
+        hash_data.extend(content_hash.to_bytes());
+
+        Ok(Hash::compute_from(&hash_data))
+    }
+}
+
+#[allow(dead_code)]
+struct BlockHeaderDenunciation {
+    public_key: PublicKey,
+    slot: Slot,
+    hash_1: Hash,
+    hash_2: Hash,
+    signature_1: Signature,
+    signature_2: Signature,
+}
+
+impl BlockHeaderDenunciation {
+    /// Compute hash for Block header content
+    fn compute_content_hash(content: &BlockHeader) -> Result<Hash, SerializeError> {
+        let mut buf = Vec::new();
+        let block_header_serializer = BlockHeaderSerializer::new();
+        block_header_serializer.serialize(content, &mut buf)?;
+        Ok(Hash::compute_from(&buf))
+    }
+
+    /// Rebuild full hash of SecuredHeader from given arguments
+    fn compute_hash_for_sig_verif(
+        public_key: &PublicKey,
+        slot: &Slot,
+        content_hash: &Hash,
+    ) -> Result<Hash, SerializeError> {
+        let mut hash_data = Vec::new();
+        let mut buf = Vec::new();
+        let de_data_serializer = DenunciationDataSerializer::new();
+
+        // Public key
+        hash_data.extend(public_key.to_bytes());
+
+        // Ser slot
+        let denunciation_data = DenunciationData::BlockHeader(*slot);
+        de_data_serializer.serialize(&denunciation_data, &mut buf)?;
+        hash_data.extend(&buf);
+        buf.clear();
+
+        // Add content hash
+        hash_data.extend(content_hash.to_bytes());
+
+        Ok(Hash::compute_from(&hash_data))
+    }
+}
+
+enum Denunciation {
+    Endorsement(EndorsementDenunciation),
+    BlockHeader(BlockHeaderDenunciation),
+}
+
+#[allow(dead_code)]
+impl Denunciation {
+    /// Check if it is a Denunciation of several endorsements
+    fn is_for_endorsement(&self) -> bool {
```

Comment:
to simplify code I wouldn't implement this helper because it is already an explicit enum